File: data-structures/stack/python/stack.py
# In Python, the most common way to implement a stack is by using a list.
# The end of the list represents the top of the stack.

def print_stack(stack):
    if not stack:
        print("Stack: [empty]")
        return

    print("Stack (bottom -> top):")
    print(stack)


def main():
    stack = []

    print("// -------------------- Creating an empty stack -------------------- //")
    print(f"Initial stack: {stack}")
    print(f"Is the stack empty? {len(stack) == 0}")

    print("\n// -------------------- push: append() -------------------- //")
    print("append(value) adds a new element to the top of the stack.\n")

    stack.append(10)
    print("After stack.append(10):")
    print_stack(stack)

    stack.append(20)
    print("After stack.append(20):")
    print_stack(stack)

    stack.append(30)
    print("After stack.append(30):")
    print_stack(stack)

    print("\n// -------------------- peek: stack[-1] -------------------- //")
    print("stack[-1] returns the top element without removing it.\n")

    top_element = stack[-1]
    print(f"Top element: {top_element}")
    print("Stack after peek:")
    print_stack(stack)

    print("\n// -------------------- size: len(stack) -------------------- //")
    print("len(stack) returns the number of elements in the stack.\n")

    print(f"Current size: {len(stack)}")

    print("\n// -------------------- is empty -------------------- //")
    print("A common way to check if the stack is empty is using 'not stack' or len(stack) == 0.\n")

    print(f"Using 'not stack': {not stack}")
    print(f"Using 'len(stack) == 0': {len(stack) == 0}")

    print("\n// -------------------- pop: pop() -------------------- //")
    print("pop() removes and returns the top element.\n")

    removed = stack.pop()
    print(f"Removed element: {removed}")
    print("Stack after pop():")
    print_stack(stack)

    removed = stack.pop()
    print(f"Removed element: {removed}")
    print("Stack after another pop():")
    print_stack(stack)

    print("\n// -------------------- Final state -------------------- //")
    print(f"Top element now: {stack[-1]}")
    print(f"Current size: {len(stack)}")
    print(f"Is the stack empty? {not stack}")


if __name__ == "__main__":
    main()

File: data-structures/stack/javascript/stack.js
// In JavaScript, the most common way to implement a stack is by using an Array.
// The end of the array represents the top of the stack.

function printStack(stack) {
    if (stack.length === 0) {
        console.log("Stack: [empty]");
        return;
    }

    console.log("Stack (bottom -> top):");
    console.log(stack);
}

function main() {
    const stack = [];

    console.log("// -------------------- Creating an empty stack -------------------- //");
    console.log("Initial stack:", stack);
    console.log("Is the stack empty?", stack.length === 0);

    console.log("\n// -------------------- push: push() -------------------- //");
    console.log("push(value) adds a new element to the top of the stack.\n");

    stack.push(10);
    console.log("After stack.push(10):");
    printStack(stack);

    stack.push(20);
    console.log("After stack.push(20):");
    printStack(stack);

    stack.push(30);
    console.log("After stack.push(30):");
    printStack(stack);

    console.log("\n// -------------------- peek: stack[stack.length - 1] -------------------- //");
    console.log("stack[stack.length - 1] returns the top element without removing it.\n");

    const topElement = stack[stack.length - 1];
    console.log("Top element:", topElement);
    console.log("Stack after peek:");
    printStack(stack);

    console.log("\n// -------------------- size: stack.length -------------------- //");
    console.log("stack.length returns the number of elements in the stack.\n");

    console.log("Current size:", stack.length);

    console.log("\n// -------------------- is empty -------------------- //");
    console.log("A common way to check if the stack is empty is using stack.length === 0.\n");

    console.log("stack.length === 0:", stack.length === 0);

    console.log("\n// -------------------- pop: pop() -------------------- //");
    console.log("pop() removes and returns the top element.\n");

    let removed = stack.pop();
    console.log("Removed element:", removed);
    console.log("Stack after pop():");
    printStack(stack);

    removed = stack.pop();
    console.log("Removed element:", removed);
    console.log("Stack after another pop():");
    printStack(stack);

    console.log("\n// -------------------- Final state -------------------- //");
    console.log("Top element now:", stack[stack.length - 1]);
    console.log("Current size:", stack.length);
    console.log("Is the stack empty?", stack.length === 0);
}

main();